src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,286 @@
+end
+
+/-- A function `f : ℍ → ℂ` is weakly modular, of weight `k ∈ ℤ` and level `Γ`, if for every matrix .
+ `γ ∈ Γ` we have `f(γ • z)= (c*z+d)^k f(z)` where `γ= ![![a, b], ![c, d]]`, and it acts on `ℍ`
+ via Moebius trainsformations. -/
+lemma weakly_modular_mem' (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) :
+  f ∈ (weakly_modular_submodule k Γ) ↔ ∀ γ : Γ, ∀ z : ℍ,
+  f (γ • z) = ((↑ₘγ 1 0 : ℝ) * z +(↑ₘγ 1 1 : ℝ))^k * f z :=
+begin
+  simp only [weakly_modular_mem],
+  split,
+  { intros h1 γ z,
+  have h3 : (f ∣[k, γ]) z = f z , by {simp_rw (h1 γ)},
+  rw [←h3, mul_comm],
+  simp only [slash_action_eq_slash, slash],
+  have h55 := zpow_neg_mul_zpow_self k (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) z),
+  simp only [upper_half_plane.denom, upper_half_plane.subgroup_to_sl_moeb, upper_half_plane.sl_moeb,
+    coe_coe, matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast] at *,
+  rw [mul_assoc, h55, ←int.coe_cast_ring_hom, ←matrix.special_linear_group.coe_matrix_coe,
+    matrix.special_linear_group.det_coe ((γ : SL(2, ℤ)) : SL(2, ℝ))],
+  simp only [of_real_one, one_zpow, mul_one], },
+  { intros hf γ,
+  simp_rw [slash_action_eq_slash],
+  ext1,
+  rw [slash, ←upper_half_plane.subgroup_moeb, (hf γ x), mul_comm],
+  have h55 := zpow_neg_mul_zpow_self k (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) x),
+  simp_rw upper_half_plane.denom at *,
+  simp only [matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix, coe_coe,
+    matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast] at h55,
+  simp only [coe_coe, matrix.special_linear_group.coe_GL_pos_coe_GL_coe_matrix,
+    matrix.map_apply, of_real_int_cast],
+  rw (matrix.special_linear_group.det_coe ((γ : SL(2, ℤ)) : SL(2, ℝ))),
+  simp only [matrix.special_linear_group.coe_matrix_coe, int.coe_cast_ring_hom, matrix.map_apply,
+    of_real_int_cast, of_real_one, one_zpow, mul_one],
+  simp_rw [← mul_assoc, h55],
+  simp },
+end
+
+lemma mul_modular (k_1 k_2 : ℤ) (Γ : subgroup SL(2,ℤ)) (f g : ℍ → ℂ)
+  (hf : f ∈ weakly_modular_submodule k_1 Γ) (hg : g ∈ weakly_modular_submodule k_2 Γ) :
+  f * g ∈ weakly_modular_submodule (k_1 + k_2) Γ :=
+begin
+  simp only [weakly_modular_mem', pi.mul_apply, coe_coe] at *,
+  intros γ z,
+  rw [(hf γ z), (hg γ z)],
+  have pown := zpow_add₀ (upper_half_plane.denom_ne_zero (γ : GL(2, ℝ)⁺) z) k_1 k_2,
+  simp only [upper_half_plane.denom, coe_fn_coe_base, ne.def,
+    matrix.general_linear_group.coe_fn_eq_coe, coe_coe] at pown,
+  rw pown,
+  ring,
+end
+
+/-- A function `f : ℍ → ℂ` is a modular form weight `k ∈ ℤ` and of level `Γ` if it is holomorphic,
+ weakly modular and bounded at infinity. -/
+structure is_modular_form_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) : Prop :=
+(hol : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (↑f : ℍ' → ℂ))
+(transf : f ∈ weakly_modular_submodule k Γ)
```

Comment:
I guess this is the main thing why you introduced `weakly_modular_submodule`? I have the feeling that a prop `is_weakly_modular` is better (it allows for dot-notation) and I did not see where you actually used the `submodule` structure.

Reply:
@mcdoll I guess the submodule structure is going to be used in setting up the graded algebra instance (?)

Reply:
I used this submodule structure to define the submodules of modular forms/ cusp forms and also for the graded ring. I could make a `is_weakly_modular` prop and then construct the weakly_modular submodule using this, if we think it'll be better?

Reply:
I generally find fewer definitions are better, it saves unfolding and re-folding.  So I would propose to keep it as it is.  @mcdoll Do you agree?

Reply:
yes, but rereading the code I wonder why not make `weakly_modular_form` and `modular_form` its own types and then it gets all the algebraic structure from the `fun_like` instance? the coercion is then a linear map from `modular_form` to `weakly_modular_form` and so on. Isn't that the standard way we define spaces of functions at the moment?

Reply:
@mcdoll This sounds vaguely familiar but I think you've worked much more with `fun_like` than me.  I'll defer to you.  Can you give a bit more of a blueprint for @CBirkbeck?  Just to check, the important thing is that there will be module instances on `space_of_cusp_forms_of_weight_and_level` and `space_of_modular_forms_of_weight_and_level` in the end; this will happen with the approach you propose, right?

Reply:
Maybe [my recent implementation of the Schwartz space](https://leanprover-community.github.io/mathlib_docs/analysis/schwartz_space.html) can be a guideline. Yes, you get the module instance from defining the coercion as a linear map and abstract nonsense (see `schwartz_map.module` in that file).

Reply:
Actually, you probably want to add classes, too. At least for weakly modular forms and modular forms. Then the modular forms and the cusp forms implement `modular_form_class` , which is an extension of `fun_like` (not directly but through `weakly_modular_form_class`). You can then implement all the cool lemmas all objects that are instances of `modular_form_class` and you get the both variants for `modular_form` and for `cusp_form`.
The class approach is used throughout the linear algebra library, [add_monoid_hom_class](https://leanprover-community.github.io/mathlib_docs/algebra/hom/group.html#add_monoid_hom_class) might be the easiest to read.
cc @j-loreaux

Reply:
Ok that `schwartz_map` example is quite nice. Let me play around with this and what [j-loreaux](https://github.com/j-loreaux) suggested and see what I can do.